### ISSUE TITLE
negative loss

### ISSUE THREAD

**Issue by styygaoo** [needylove/OrdinalEntropy#4]
Hi,

Thanks for your great work. I am trying to apply your ordinal entropy loss to my network. Specifically, I just use the diverse part of the ordinal entropy loss since my model is designed for depth estimation. Surprisingly, I always get a negative value of the diverse part. It seems that the performance of my network is worse and worse by applying ordinal entropy loss since I can see the loss value was getting bigger and bigger.

Could you give me some suggestions about my problem?

Thanks in advance!

below is the loss function used:

def ordinalentropy_loss(features, gt,  mask=None):
    """
    Features: a certain layer's features
    gt: pixel-wise ground truth values, in depth estimation, gt.size()= n, h, w
    mask: In case values of some pixels do not exist. For depth estimation, there are some pixels lack the ground truth values
    """
    # gc.collect()
    # torch.cuda.empty_cache()

    f_n, f_c, f_h, f_w = features.size()

    features = F.interpolate(features, size=[f_h // 4, f_w // 4], mode='nearest')
    features = features.permute(0, 2, 3, 1)  # n, h, w, c
    features = torch.flatten(features, start_dim=1, end_dim=2)

    gt = F.interpolate(gt, size=[f_h // 4, f_w // 4], mode='nearest')

    loss = 0

    for i in range(f_n):
        """
        mask pixels that without valid values
        """
        # print("{:.3f}MB allocated".format(torch.cuda.memory_allocated() / 1024 ** 2))
        torch.cuda.empty_cache()
        _gt = gt[i,:].view(-1)
        _mask = _gt > 0.001
        _mask = _mask.to(torch.bool)
        _gt = _gt[_mask]
        _features = features[i,:]
        _features = _features[_mask,:]

        """
        diverse part
        """
        u_value, u_index, u_counts = torch.unique(_gt, return_inverse=True, return_counts =True)
        center_f = torch.zeros([len(u_value), f_c]).cuda()
        center_f.index_add_(0, u_index, _features)
        u_counts = u_counts.unsqueeze(1)
        center_f = center_f / u_counts

        p = F.normalize(center_f, dim=1)
        # print("{:.3f}MB allocated".format(torch.cuda.memory_allocated() / 1024 ** 2))
        _distance = euclidean_dist(p, p)
        _distance = up_triu(_distance)

        u_value = u_value.unsqueeze(1)
        _weight = euclidean_dist(u_value, u_value)
        _weight = up_triu(_weight)
        _max = torch.max(_weight)
        _min = torch.min(_weight)
        _weight = ((_weight - _min) / _max)

        _distance = _distance * _weight

        _entropy = torch.mean(_distance)
        loss = loss - _entropy
        # print(loss)
        """
        tightness part
        """
        # _features = F.normalize(_features, dim=1)
        # _features_center = p[u_index, :]
        # _features = _features - _features_center
        # _features = _features.pow(2)
        # _tightness = torch.sum(_features, dim=1)
        # _mask = _tightness > 0
        # _tightness = _tightness[_mask]
        #
        # _tightness = torch.sqrt(_tightness)
        # _tightness = torch.mean(_tightness)
        #
        # loss = loss + _tightness
    gc.collect()
    torch.cuda.empty_cache()
    return loss/ f_n

**Comment by needylove** [comment]
Hi, we maximize the diversity part by minimizing the negative diversity part. Thus, you should obtain a negative value (i.e., loss = loss - _entropy, here _entropy>0, and thus loss<0). You can try different weights for this term for better performance. Besides, although a higher diversity is beneficial for the representation ability, it may have a negative impact on the generalization ability, a related work can be found here: https://openreview.net/forum?id=OeNcnlQPRz.

**Comment by styygaoo** [comment]
Hi,
Thank you very much for your reply. It is very helpful. I will have a close look at the paper you mentioned.
Best regards,

**Comment by needylove** [comment]
You are welcome. Please let us know if you have any questions
